pkg/trace/writer/newwriter/sender.go

```diff
@@ -0,0 +1,382 @@
+
+// sender books payloads for being sent to a given URL by a given client. It uses a
+// size-limited retry queue with a backoff mechanism in case of retriable errors.
+type sender struct {
+	cfg  *senderConfig
+	host string
+
+	wg     sync.WaitGroup // waits for all uploads
+	climit chan struct{}  // acts as a semaphore for limiting concurrent connections
+
+	mu        sync.Mutex  // guards below fields
+	list      *list.List  // send queue
+	size      int         // size of send queue
+	scheduled bool        // reports if a flush is scheduled
+	attempt   int         // retry attempt coming up next
+	timer     *time.Timer // timed flush (triggered by backoff)
+}
+
+// senderConfig specifies the configuration for the sender.
+type senderConfig struct {
+	// client specifies the HTTP client to use when sending requests.
+	client *http.Client
+	// url specifies the URL to send requests too.
+	url *url.URL
+	// apiKey specifies the Datadog API key to use.
+	apiKey string
+	// maxConns specifies the maximum number of allowed concurrent ougoing
+	// connections.
+	maxConns int
+	// recorder specifies the eventRecorder to use when reporting events occurring
+	// in the sender.
+	recorder eventRecorder
+}
+
+// newSender creates a new sender, which uses the given HTTP client, URL and API key
+// to make requests using incoming payloads. The limit of outgoing concurrent connections
+// is defined by climit.
+func newSender(cfg *senderConfig) *sender {
+	return &sender{
+		cfg:    cfg,
+		climit: make(chan struct{}, cfg.maxConns),
+		list:   list.New(),
+	}
+}
+
+// Push pushes p onto the sender, to be written to the destination.
+func (q *sender) Push(p *payload) {
+	q.mu.Lock()
+	defer q.mu.Unlock()
+
+	// enqueue the payload
+	q.enqueueLocked(p)
+
+	if !q.scheduled {
+		// no flush is scheduled; start one
+		q.scheduled = true
+		q.wg.Add(1)
+		go q.flush()
+	}
+}
+
+// flush drains and sends the entire queue. If anything comes in while flushing
+// or if some of the payloads fail to send as retriable, further follow up flushes
+// are scheduled.
+func (q *sender) flush() {
+	startTime := time.Now()
+
+	// we drain the queue, which is a blocking operation, meaning that
+	// new payloads are stopped from coming in while this happens; but
+	// it is a fast.
+	payloads := q.drainQueue()
+
+	// we send the payloads we've retrieved; while we do this, more payloads
+	// can join the list.
+	done, retries := q.sendPayloads(payloads)
+
+	if done > 0 {
+		q.recordEvent(eventTypeFlushed, &eventData{
+			count:    int(done),
+			duration: time.Since(startTime),
+		})
+	}
+
+	// we reassess the state of the list and check if further flushes need
+	// to be scheduled
+	q.mu.Lock()
+	defer q.mu.Unlock()
+	defer q.wg.Done()
+
+	if q.list.Len() == 0 {
+		// the list is empty; no further flushing needs to be triggered
+		q.attempt = 0
+		q.scheduled = false
+		return
+	}
+	// the list is not empty
+	q.scheduled = true
+	if retries > 0 {
```

Comment:
Since this is already treated as a special case (early return), I think it makes more sense to be the first special case to treat instead of `q.list.Len() == 0` (similar to how you usually handle the error scenario close to the place that returns the `err`).

Reply:
Oh yeah. That totally looks easier to read. Great suggestion.